faf-commons-data/src/main/java/com/faforever/commons/replay/RawReplayDataReader.java

```diff
@@ -0,0 +1,84 @@
+package com.faforever.commons.replay;
+
+import com.google.common.io.BaseEncoding;
+import com.google.common.io.ByteArrayDataInput;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+
+/**
+ * Utility class for unpacking replay data.
+ */
+public class RawReplayDataReader {
+  private enum CompressionType {
+    LEGACY,
+    LEGACY_ZSTD,
+  }
+
+  private static byte[] uncompress(byte[] data, int start, int end, CompressionType ctype) throws IOException, InvalidReplayException {
+    ByteArrayDataInput i;
+    switch (ctype) {
+      case LEGACY:
+        /* FIXME data copy */
+        String b64encodedData = new String(data, start, end - start, "UTF-8");
+        byte decodedData[];
+        try {
+          decodedData = BaseEncoding.base64().decode(b64encodedData);
+        } catch (IllegalArgumentException e) {
+          throw new InvalidReplayException("Replay data is not a valid base64 string");
+        }
+        return QtCompress.qUncompress(decodedData);
+      case LEGACY_ZSTD:
+        return ZstdUncompress.zstdUncompress(data, start, end);
+      default:  /* Never happens anyway */
+        throw new IOException("Unknown replay compression format");
+    }
+  }
+
+  private static CompressionType getCompressionType(byte[] data, int start, int end) throws InvalidReplayException, IOException {
+    JsonObject headerObject;
+    try {
+      Gson parser = new Gson();
+      headerObject = parser.fromJson(
+        new String(Arrays.copyOfRange(data, start, end), "UTF-8"),
```

Comment:
Never manually write Strings for "default" stuff. Use `StandardCharsets.UTF_8` instead.